### ISSUE TITLE
CA1066 false positive

### ISSUE THREAD

**Issue by froufrou121** [dotnet/roslyn-analyzers#2079]
#### Analyzer package

Example: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

 v2.6.3

#### Diagnostic ID

Example: [CA1716](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1716-identifiers-should-not-match-keywords)

#### Repro steps

```csharp
namespace RangeCs8
{
    using System;
    using System.Globalization;

#nullable enable

    public struct Range<T> : IEquatable<Range<T>>
        where T : struct, IComparable<T>
    {
        private const string NegativeInfinity = "-∞";

        private const string PositiveInfinity = "+∞";

        public Range(T lowerBound, T upperBound)
        {
            if (lowerBound.CompareTo(upperBound) > 0)
            {
                throw new ArgumentException("lower bound must be < upper bound.");
            }

            this.LowerBound = lowerBound;
            this.UpperBound = upperBound;
        }

        public T LowerBound { get; }

        public T UpperBound { get; }

        private static T MinValue => GetFieldValue("MinValue");

        private static T MaxValue => GetFieldValue("MaxValue");

        public static bool operator ==(Range<T> range1, Range<T> range2) => range1.Equals(range2);

        public static bool operator !=(Range<T> range1, Range<T> range2) => !range1.Equals(range2);

        public override bool Equals(object obj) => obj is Range<T> && this.Equals((Range<T>)obj);

        public bool Equals(Range<T> other)
        {
            return this.LowerBound.CompareTo(other.LowerBound) == 0 &&
                this.UpperBound.CompareTo(other.UpperBound) == 0;
        }

        public override int GetHashCode() => this.LowerBound.GetHashCode() ^ this.UpperBound.GetHashCode();

        public override string ToString()
        {
            return string.Format(
                CultureInfo.CurrentCulture,
                "[{0}, {1}[",
                this.LowerBound.CompareTo(MinValue) == 0 ? NegativeInfinity : (object)this.LowerBound,
                this.UpperBound.CompareTo(MaxValue) == 0 ? PositiveInfinity : (object)this.UpperBound);
        }

        public bool IsInRange(T value)
        {
            return value.CompareTo(this.LowerBound) >= 0 && value.CompareTo(this.UpperBound) < 0;
        }

        private static T GetFieldValue(string fieldName)
        {
            var fi = typeof(T).GetField(fieldName);
            if (fi != null)
            {
                try
                {
                    return (T)fi.GetValue(null);
                }
                catch (InvalidCastException)
                {
                    return default;
                }
            }

            return default;
        }
    }
}
```
#### Expected behavior

no CA1066 warning: public bool Equals(Range<T> other) is defined

#### Actual behavior

2>C:\Users\frous\source\repos\RangeCs8\RangeNet\Class1.cs(8,19,8,24): warning CA1066: Implement IEquatable when overriding Object.Equals

**Comment by mavasani** [comment]
Unable to repro on the latest FxCop analyzers package: 2.9.4-beta1.final

**Comment by froufrou121** [comment]
I have updated to 2.9.4-beta1.final but I still get warning CA1066 with the supplied code.

**Comment by froufrou121** [comment]
[RangeCs8.zip](https://github.com/dotnet/roslyn-analyzers/files/3349020/RangeCs8.zip)
You can try with my solution.

**Comment by mavasani** [comment]
Interesting, this only repros if `#nullable enable` is used. @333fred @jasonmalinowski It seems like a some breaking API change in presence of nullability?

**Comment by mavasani** [comment]
@333fred This seems like a breaking change (bug?) in `INamedTypeSymbol.Construct` API in presence of nullability. Analyzer code that breaks:

https://github.com/dotnet/roslyn-analyzers/blob/8984fd672f4d1ad6ce05bebe048e841d1f99d4b4/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/EquatableAnalyzer.cs#L75-L79

![image](https://user-images.githubusercontent.com/10605811/60511706-7214e200-9c87-11e9-838f-2f87501a7f9d.png)

It seems that the type argument `T` from `constructedEquatable` has nullable annotation `Oblivious`, but the one from `namedType.AllInterfaces` has nullable annotation `NotAnnotated`

**Comment by jasonmalinowski** [comment]
So this is trying to say "do I implement IEquatable of myself?" If so, what may be necessary here is .Equals() should be ignoring nullability unless you opted into it, but that's not the current compiler API behavior. https://github.com/dotnet/roslyn/issues/35933 is tracking fixing that.

**Comment by mavasani** [comment]
@jasonmalinowski Thank you!

@froufrou121 - I am going to close this as external. You will need to wait for a VS/compiler version with has the fix for dotnet/roslyn#35933
